### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by sjanssen2**
* basically a complete re-write of the server. I am only listing the new features here:
    * directly jump to the interactive page
    * top banner that lists the user selected instance as a constant reminder of what shall be executed: "Your instance: ..."
    * algebras and product operator choice is dynamically added when user makes a choice other then empty for the rightmost algebra. Max is currently set to 5 algebras (change via cafe.py)
    * IF "example inputs: " is found as a comment in the *.gap source file, these strings are used to pre-populate input fields and to giva a hint.
    * results are split into following fields and nicely rendered via highlight.js
      - command
      - stdout
      - stderr (if any)
      - used cache directory, runtime and max RSS
    * source code is listed as an accordion to the page
    * source "download" with txt mimetype
    * IF top comment exists in *.gap file, a "Problem Description" is given as an accordion

### CODE REVIEW

File: Bellmansgap.py [hunk 1]
```diff
@@ -1,34 +1,12 @@
...
-    os.symlink("../" + PREFIX_GAPUSERSOURCES + "Resources", "static/Resources")
-
-# user submission leads to compilation and execution of new algera products
-# if the user re-submits the same algebra product (also called instance) it
-# does not need to be re-computed, therefore we are using a cache. JUST this
-# instance with user inputs have to be run.
-PREFIX_CACHE = "DOCKER/bcafe_cache/"
 
 app = Flask(__name__)
 app.secret_key = "xasdqfghuioiuwqenjdcbjhawbuomcujeq1217846421kopNSJJGWmc8u29"
```
Comment (Anna-Rehm) [inline Bellmansgap.py:12]: The secret key should not be open to public as it poses a security threat.
Rather, import from another file and hide that file with .gitignore

Comment (sjanssen2) [reply Bellmansgap.py:12]: absolutely right. We should also not execute the server in debug mode: [`app.run(debug=True)`](https://github.com/jlab/bellmanscafe/blob/03534e810d18756ae535dc35ad32f8ab15516024/Bellmansgap.py#L88)

File: static/css/style.css [hunk 2]
```diff
@@ -0,0 +1,194 @@
+:root {
+  --jlab_blue: #0068b4;      /* was: #AF6557 */
+  --jlab_lightblue: #89c7f4; /* was: #B8E1A3 */
```
Comment (Anna-Rehm) [inline static/css/style.css:3]: A bit too light as a bg colour for white script. But this is an aesthetic issue and can be changed down the line :)

**Review by Anna-Rehm — CHANGES_REQUESTED** [review]
Issue with secret key should be solved before merging, already addressed [here](https://github.com/jlab/bellmanscafe/pull/54#discussion_r1452533442)

File: Bellmansgap.py (hunk 1 above)
Comment (sjanssen2) [reply Bellmansgap.py]: key is moved into `gunicorn_bellmanscafe_conf.py` and changed on the actual serving instance

**Review by Anna-Rehm — APPROVED** [review]

File: static/css/style.css (hunk 2 above)
Comment (kmaibach) [reply static/css/style.css:3]: the complementary colour of "jlab blue" would be `#b44c00`
which is a darker orange.

could work, but could also look horrible :D 
I tried it on the website but I am not sure what to think of the combination.

Alternatively `#0085e7` might look good being a lighter shade of "jlab blue"

**Review by kmaibach — APPROVED** [review]
I might have missed this but what about all the files generated during one run (the compiled files, the tikz LaTeX file and so on). Are they stored on the server somewhere? Are they just cached and deleted? 
Do we need to think about a good solution or is this already taken care of?

File: Bellmansgap.py (hunk 1 above)
Comment (sjanssen2) [reply Bellmansgap.py]: yes, I am using a new key know - only known in the VM

**Comment by sjanssen2** [comment]
> I might have missed this but what about all the files generated during one run (the compiled files, the tikz LaTeX file and so on). Are they stored on the server somewhere? Are they just cached and deleted? Do we need to think about a good solution or is this already taken care of?

These files are stored in https://github.com/jlab/bellmanscafe/blob/8b153feb277d7f2ee66297d880f288502ed2a576/gunicorn.conf.py#L44 and used as a cache. The cache directory depends on a) the `gapc --version` b) a hash sum of the *.gap file and c) another hash sum of the user track input. Updating one of these components automatically leads to other directories.
Currently, I am not limiting the total size of the cache as I do not assume too many users. We can add this once we hit the disk quota ;-)

**Comment by sjanssen2** [comment]
> Nice work! Server is looking sharp.

thanks :-)

> I would consider adding the python package markdown to the requirements in the README.txt Maybe even add Flask. gunicorn and markdown to a separate requirements.txt

I feel we need to completely rewrite the readme.md :-/

> If we do this than the Dockerfile would also need to be adjusted.

Ideally, we would "dockerize" the server, such that the BCF can easily maintain this server for us. I have not yet found the time to complete this task.
 
> If needed the elmamun example can be sharpened using AI service https://neural.love/image-sharpen

I don't have any credits but would love to see the result.

**Pull request merged**